kwasm-lemmas.md

```diff
@@ -20,13 +20,156 @@ Basic logic
 Basic arithmetic
 ----------------
 
+// TODO: Upstream the lemmas in this section into K.
+
+### Modular Arithmetic
+
+Z3 is slow and unreliable in reasoning about modular arithmetic.
+Therefore we want to make structural simplifications wherever possible.
+
+#### Rules for Expressions With Only Modulus
+
 ```k
     rule X modInt N => X
       requires 0 <=Int X
        andBool X  <Int N
       [simplification]
 ```
 
+`modInt` selects the least non-negative representative of a congruence class modulo `N`.
+
+```k
+    rule (X modInt M) modInt N => X modInt M
+      requires M >Int 0
+       andBool M <=Int N
+      [simplification]
+```
+
+Proof:
+
+```
+Since 0 <= x mod m < m <= n, (x mod m) mod n = x mod m
+```
+
+```k
+    rule (X modInt M) modInt N => X modInt N
+      requires M >Int 0
+       andBool N >Int 0
+       andBool M modInt N ==Int 0
+      [simplification]
+```
+
+Proof:
+
+```
+Assume m = n * k for some k > 0.
+x = m * q + r, for a unique q and r s.t. 0 <= r < m
+(x mod m) mod n
+ = r mod n
+ = n * ( k * q) + r mod n
+ = m * q + r mod n
+ = x mod n
+```
+
+#### Modulus Over Addition and Division
+
+```k
+    rule (X *Int M +Int Y) modInt N => Y modInt N   requires M modInt N ==Int 0 [simplification]
+    rule (Y +Int X *Int M) modInt N => Y modInt N   requires M modInt N ==Int 0 [simplification]
+```
+
+Proof:
+
+```
+Assume m = k * n for some k (side condition).
+x * m + y mod n = x * (k * n) + y mod n = y mod n
+```
+
+```k
+    rule ((X modInt M) +Int Y) modInt N => (X +Int Y) modInt N   requires M modInt N ==Int 0 [simplification]
+    rule (X +Int (Y modInt M)) modInt N => (X +Int Y) modInt N   requires M modInt N ==Int 0 [simplification]
+```
+
+Proof:
+
+```
+Assume m = l * n
+x = k * m + r, r < m
+x mod m + y = r + y
+(x + y) mod n
+  = k * m + r + y mod n
+  = k * l * n + r + y mod n
+  = r + y mod n
+  = (x mod m + y) mod n
+```
+
+#### Bit Shifting
+
+We want Z3 to understand what a bit-shift is.
+
+```k
+    rule ( X >>Int N)          => 0 requires X <Int 2 ^Int N [simplification]
+    rule ( X <<Int N) modInt M => 0 requires M <Int 2 ^Int N [simplification]
+
+    rule (X >>Int N) >>Int M => X >>Int (N +Int M) [simplification]
+    rule (X <<Int N) <<Int M => X <<Int (N +Int M) [simplification]
+
+    // The Haskell and Java backend accept negative shifts (the LLVM and OCaml backends do not).
+    // So removing the side conditions and keeping one of each rule here could give faster symbolic execution.
+    rule (X <<Int N) >>Int M => X <<Int (N -Int M)   requires          N >=Int M  [simplification]
+    rule (X <<Int N) >>Int M => X >>Int (M -Int N)   requires notBool (N >=Int M) [simplification]
+    rule (X >>Int N) <<Int M => X >>Int (N -Int M)   requires          N >=Int M  [simplification]
+    rule (X >>Int N) <<Int M => X <<Int (M -Int N)   requires notBool (N >=Int M) [simplification]
```

Comment:
Last 2 are unsound

Reply:
Can be expressed as a "blanking operation"
Shifting 8 bites right and then 8 bits left is blanking the leat 8 bits.